block-layer/graphsync/known_extensions.md

```diff
@@ -0,0 +1,63 @@
+# Graphsync: Known Extensions
+
+### Excluded CIDs
+
+Extension Name: `graphsync/exclude-cids`
+
+What it does:
+
+Often a node may know ahead of time that it has some of the blocks needed to match a selector query in its local store already. Some reasons this might occur include:
+
+- a previous request was interrupted
+- a previous request for a subset of the requested selector was already completed
+
+How it works:
+
+When a requestor sends a request, it should send a CBOR encoded IPLD Node that is a list of CIDs it already has as the value for this extension.
+
+The IPLD schema is as follows:
+
+```ipldsch
+type ExcludedCids [Cid]
+```
+
+The responder node, if it supports the extension, will not send those blocks back in the response. It does not send a value back in the response for this extension.
+
+### Response Metadata
```

Comment:
* Why do we need this extension? I _should_ be able to derive this information locally at no cost (just execute the selector). I can't _not_ execute the selector for security reasons.
* Are we planning on _relying_ on this extension? If so, we _really_ should be baking it into the protocol.

Reply:
> The proposal for the initial go implementation, in order to optimize memory usage, is to instead have the server send blocks as soon as it finds them, as well as information about what blocks are not available as soon as it knows. Then the client will verify and return blocks to the graphsync caller as soon as it receives them.

(from https://github.com/ipld/specs/pull/111#issue-269750451)

This worries me. We _can't_ return blocks to the caller until _we've_ verified that they're a part of the query. We can't just verify that the blocks match the CIDs reported by the remote side.

Reply:
To be clear, the response metadata is already part of the go-graphsync implementation.

The comment above is not clear, and expresses intent poorly. The word "verify" in this case means verify that they are part of the query.

The most useful thing that response metadata does is tell the client about CIDs that it doesn't have. By knowing that CID's aren't present, and won't be incoming, the client can know immediately it can progress a it's local selector query and return an error for that part of the tree.

I can really work on rewriting this.

Reply:
> The word "verify" in this case means verify that they are part of the query.

The client can't actually do that without the blocks, right?

> The most useful thing that response metadata does is tell the client about CIDs that it doesn't have. 

_That_ actually sounds pretty important. Should that not be a mandatory field? I guess it's _technically_ not required for correctness but I can't imagine an DoS-proof graphsync implementation that doesn't need this. That is, I'd expect a secure graphsync implementation to disconnect from peers that send over N (100?) blocks while we can't make progress in the query.